Client/core/CDiscordManager.cpp

```diff
@@ -0,0 +1,345 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        Client/core/CDiscordManager.cpp
+ *
+ *  Multi Theft Auto is available from http://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+
+#include "StdInc.h"
+#include "SharedUtil.Thread.h"
+#include <chrono>
+
+CDiscordManager::CDiscordManager() : m_discordCore(nullptr), m_bSuicide(false), m_waitingForServerName(false), m_storedActivity{}
+{
+    Reconnect(true); // Try to interact with discord on construction
+    m_pThread = new CThreadHandle(CDiscordManager::DiscordThread, this);
+
+    m_storedActivity.GetAssets().SetLargeImage("multi-theft-auto"); // Always thing
+    m_storedActivity.GetAssets().SetLargeText("Playing MTA:SA");
+}
+
+CDiscordManager::~CDiscordManager()
+{
+    m_bSuicide = true;
+    int iTries = 0;
+    while (iTries++ < 400 && m_bSuicide) // Wait maximum of 2 sec on this
+        Sleep(5);
+    if (m_bSuicide) m_pThread->Cancel(); // Kill it anyway
+    delete m_pThread;
+    SAFE_DELETE(m_discordCore);
+}
+
+// Establish connection with discord
+void CDiscordManager::Reconnect(bool bOnInitialization)
+{
+    bool discordRichPresence;
+    CVARS_GET("discord_rich_presence", discordRichPresence);
+    if (!discordRichPresence) return; // Disabled
+
+    m_threadSafety.lock();
```

Comment:
Just use a lock guard here. `lock()` and `unlock()` are very easy to miss and could be moved around accidentally.

Reply:
I can't unless I move that line to another scope, it will result in circular mutex lock.

Reply:
You could create a new scope.

E.g.:
```cpp
// ...
if (!discordRichPresence) return; // Disabled

{
  std::lock_guard<std::mutex>{m_threadSafety};
  discord::Result res = discord::Core::Create(DISCORD_CLIENT_ID, DiscordCreateFlags_NoRequireDiscord, &m_discordCore);
}
// ...
```

Reply:
So another scope it is, alright.